cs/test/core/DeserializerControlsTests.cs

```diff
@@ -0,0 +1,150 @@
+        }
+
+        // Restore the default settings at the start and end of each test
+        [SetUp]
+        [TearDown]
+        public void RestoreDefaults()
+        {
+            DeserializerControls.Active = DeserializerControls.Default;
+        }
+
+        void RoundtripObject<T>(T obj)
+        {
+            // SimpleBinary
+            {
+                var output = new OutputBuffer();
+                var writer = new SimpleBinaryWriter<OutputBuffer>(output);
+                Serialize.To(writer, obj);
+
+                var input = new InputBuffer(output.Data);
+                var reader = new SimpleBinaryReader<InputBuffer>(input);
+                T obj2 = Deserialize<T>.From(reader);
+
+                Assert.True(Comparer.Equal<T>(obj, obj2));
+            }
+
+            // CompactBinary
+            {
+                var output = new OutputBuffer();
+                var writer = new CompactBinaryWriter<OutputBuffer>(output);
+                Serialize.To(writer, obj);
+
+                var input = new InputBuffer(output.Data);
+                var reader = new CompactBinaryReader<InputBuffer>(input);
+                T obj2 = new Deserializer<CompactBinaryReader<InputBuffer>>(typeof(T)).Deserialize<T>(reader);
+
+                Assert.True(Comparer.Equal<T>(obj, obj2));
+            }
+
+            // SimpleXML
+            {
+                var xmlString = new StringBuilder();
+                var xmlWriter = new SimpleXmlWriter(XmlWriter.Create(xmlString));
+
+                Serialize.To(xmlWriter, obj);
+                xmlWriter.Flush();
+
+                var reader = new SimpleXmlReader(XmlReader.Create(new StringReader(xmlString.ToString())));
+                T obj2 = new Deserializer<SimpleXmlReader>(typeof(T)).Deserialize<T>(reader);
+
+                Assert.True(Comparer.Equal<T>(obj, obj2));
+            }
+        }
+
+        void RoundtripTests()
+        {
+            var containers = new SimpleContainers();
+            for (int i = 0; i < 30; i++)
+            {
+                containers.strings.Add(i.ToString());
+                containers.numbers.Add(i, i.ToString());
+            }
+
+            RoundtripObject<SimpleContainers>(containers);
+
+            var blobs = new StructWithBlobs();
+            blobs.b = new ArraySegment<byte>(blobData, 0, 30);
+            blobs.lb.Add(new ArraySegment<byte>(blobData, 0, 30));
+            blobs.lb.Add(new ArraySegment<byte>(blobData, 1, 30));
+
+            RoundtripObject<StructWithBlobs>(blobs);
+        }
+
+        [Test]
+        public void DeserializerControls_InvalidArguments()
+        {
+            Assert.Throws<ArgumentOutOfRangeException>(() => DeserializerControls.Active.MaxPreallocatedContainerElements = -1);
+            Assert.Throws<ArgumentOutOfRangeException>(() => DeserializerControls.Active.MaxPreallocatedBlobBytes = -1);
+        }
+
+        [Test]
+        public void DeserializerControls_InvalidData()
```

Comment:
Take a look at the [Mono run](https://travis-ci.org/Microsoft/bond/jobs/191072880). The tests are failing with different exceptions there.